src/sqlfluff/dialects/dialect_snowflake.py

```diff
@@ -105,7 +107,7 @@
         type="variable",
     ),
     ReferencedVariableNameSegment=RegexParser(
-        r"(\$|\&)[A-Z][A-Z0-9_]*",
+        r'(?:\$[A-Z][A-Z0-9_]*)|(?:[A-Z0-9_\."]*&{[A-Z0-9_]*}[A-Z0-9_]*)|(?:&[A-Z0-9_\."]*)',
```

Comment:
For some reason `(?:[\w\."]*&{[A-Z0-9_]*}[\w]*)` does not work, while `(?:[A-Z0-9_\."]*&{[A-Z0-9_]*}[A-Z0-9_]*)` does. Does anybody know why?

Reply:
Those should be identical. Can you define "does not work"? Have you been able to reproduce it on https://regex101.com/?

Reply:
Sorry, you are right, "does not work" is rarely helpful.
When I use `r'(?:\$[A-Z][A-Z0-9_]*)|(?:[A-Z0-9_\."]*&{[A-Z0-9_]*}[A-Z0-9_]*)|(?:&[A-Z0-9_\."]*)',`, then `USE WAREHOUSE &WAREHOUSE;` parses alright. If I use `r'(?:\$[A-Z][A-Z0-9_]*)|(?:[\w\."]*&{[A-Z0-9_]*}[A-Z0-9_]*)|(?:&[\w\."]*)',`, the same piece of SQL returns a _Found unparsable section_ error.
I cannot reproduce this on regex101.com - over there both patterns produce the desired match `&WAREHOUSE`.

Reply:
This is a bug - not in your regex - but in the RegexParser that is being used. RegexParser inherits from StringParser, which calls this in its __init__ (which is called in the RegexParser)
```
self.template = template.upper()
```
So, `\w` becomes `\W`, which matches every character that is not a word character. Basically `^[\w]`